src/txmempool.cpp

```diff
@@ -163,6 +163,17 @@ util::Result<CTxMemPool::setEntries> CTxMemPool::CalculateAncestorsAndCheckLimit
     int64_t totalSizeWithAncestors = entry_size;
     setEntries ancestors;
 
+    // Package itself is busting mempool limits; should be rejected even if no staged_ancestors exist
+    if (entry_count > static_cast<uint64_t>(limits.ancestor_count)) {
+        return util::Error{Untranslated(strprintf("too many ancestors in package %u [limit: %u]", entry_count, limits.ancestor_count))};
+    } else if (entry_count > static_cast<uint64_t>(limits.descendant_count)) {
+        return util::Error{Untranslated(strprintf("too many descendants in package %u [limit: %u]", entry_count, limits.descendant_count))};
+    } else if (entry_size > limits.ancestor_size_vbytes) {
+        return util::Error{Untranslated(strprintf("exceeds ancestor size limit %u [limit: %u]", entry_size, limits.ancestor_size_vbytes))};
+    } else if (entry_size > limits.descendant_size_vbytes) {
+        return util::Error{Untranslated(strprintf("exceeds descendant size limit for tx %u [limit: %u]", entry_size, limits.descendant_size_vbytes))};
```

Comment:
8a8a549a330ef3de99a6a3d7c150a6e2ab5ff11e

Why is this in `CalculateAncestorsAndCheckLimits` instead of `CheckPackageLimits`? Would be faster to check this before calculating `staged_ancestors`.

Also I wonder if this could be a confusing error string due to how we're checking. e.g. descendant limit < ancestor limit, so a child-with-24-parents package hits "too many descendants" even if everyone has a descendant count of 1 or 2.

Perhaps more clear is "package count exceeds ancestor count limit," "package vsize exceeds ancestor count limit," etc.

Reply:
it's not a "Package limit" per se, since it's specifically mempool chain limits?

Reply:
done